Mlem/API/APIClient.swift

```diff
@@ -0,0 +1,94 @@
+//
+//  APIClient.swift
+//  Mlem
+//
+//  Created by Nicholas Lawson on 04/06/2023.
+//
+
+import Foundation
+
+extension JSONDecoder {
+    static var defaultDecoder: JSONDecoder {
+        let decoder = JSONDecoder()
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+        
+        let formatter = DateFormatter()
+        
+        decoder.dateDecodingStrategy = .custom({ decoder in
+            let container = try decoder.singleValueContainer()
+            let string = try container.decode(String.self)
+            
+            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS"
+            if let date = formatter.date(from: string) {
+                return date
+            }
+            
+            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
+            if let date = formatter.date(from: string) {
+                return date
+            }
+            
+            // TODO: is there somewhere that covers all the formats lemmy instances may return?
+            // for now we'll return the current date until we know what other formats will
+            // be encountered.
+            return .now
+        })
+        return decoder
+    }
+}
+
+enum HTTPMethod {
+    case get
```

Comment:
There's three methods; GET, POST, and PUT

POST and PUT work the same in that they just send a JSON body, but GET serialises everything to its query parameters

Maybe its better to generalise data into this field, and then remove the query items bit on the APIRequest, because it'll never get used for PUT/POST, and vice versa, json body data will never be used for GET

Reply:
I was thinking for POST/PUT to accept any Data, but maybe then GET would accept a `[String: String]` then instead?

Reply:
addressed by dbfe9740835ee1db247427226641e3617925b840